platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

```diff
@@ -0,0 +1,812 @@
+7. **EC-Key Utils** is a utility module that enables the node operator to generate a bootstrapping public/private key pair.
+
+### Module organization and repositories
+![img_6.png](img_6.svg)
+1. **hedera-cryptography**: This is a separate repository for hosting cryptography-related libraries.
+   It is necessary to facilitate our build process, which includes Rust libraries. It also provides independent release cycles between consensus node code and block node code.
+2. **swirlds-native-support**: Gradle module that enables loading into memory compiled native libraries so they can be used with JNI.
+3. **swirlds-crypto-tss**: Gradle module for the TSS Library.
+   This library's only client is the consensus node, so it will be in the hedera-services repository, under the `platform-sdk` folder.
+4. **swirlds-crypto-signatures**: Gradle module for the Bilinear Pairings Signature Library.
+5. **swirlds-crypto-pairings-api**: Gradle module for the Bilinear Pairings API. Minimizes the impact of adding or removing implementations.
+6. **swirlds-crypto-altbn128**: Gradle module that will implement the Bilinear Pairings API using alt-bn128 elliptic curve.
+   That curve has been chosen due to EVM support. The arkworks rust library will provide the underlying cryptography implementation.
+   The module will include Java and Rust codes that will be compiled for all possible system architectures and distributed in a jar under a predefined structure.
+
+### Handling of multilanguage modules and native code distribution
+The code provided by this solution will include multilanguage modules (rust + java). Rust code needs to be compiled into binary libraries and accessed through JNI.
+Given the non-GA state of Project Panama, this proposal does not consider it, but it has not been discarded for future change proposals.
+
+There are two possible ways of loading libraries and accessing them through JNI
+1) Libraries are installed in the SO and referenced through the classpath.
+2) Distributed with the application jars, unzipped, and loaded in runtime.
+
+We want to ensure that dependent software does not require the installation of any additional dependencies other than those distributed in the jars, so we are choosing option 2.
+
+In multilanguage projects, Rust code will be compiled first using `Cargo` into a binary library and then packaged in the same jar.
+While developing and on a developer’s machine, Rust code will be built for the architecture executing the build and packaged in the library’s jar, accessing the JNI wrapping code.
+Once the code is merged to develop, the CI/CD pipeline compiles the Rust code into a binary library for each of the multiple supported platforms, packages it into a jar, and publishes it to Maven.
+The dependency from the Maven repo will contain the Java code and the binary library cross-compiled for all supported architectures.
```

Comment:
I have a question about this item.

Would it make better sense (and be more secure) to only include one architecture per jar, and select the relevant jar for a given architecture?

We could then produce OCI images for each supported architecture containing only the native code for that architecture, and thus avoid distributing incompatible code that might have potential for issues in the system.

Reply:
This build process was defined with @nathanklick.
The benefit is that you maintain the "build once, run everywhere" property. (where everywhere in this case is just supported architectures, as you mentioned above)

The risk you mention is mitigated by the fact that the code will make the right decision regarding the library to load and only unpackage the right binary for the architecture in execution.

Of course, if somebody gains access to the node, unpackages the content of this jar, and executes the wrong file, bad things will happen. Is there any other scenario you think of where the distributed binary can create a problem if it remains packaged?

@hendrikebbers had a similar proposal as yours ragarding this topic.
I guess there are pros and cons to both strategies.

I think we should go with the one defined here, but if you still strongly oppose it, I can arrange a meeting with all interested parties.

Reply:
It's not a strong opposition.  More of something to bring up with our product security experts. It may be possible, in rare circumstances, to make running code try to load a bad library (that's been done to cause failures in other systems, if I recall correctly).

Run-anywhere from a single jar is less valuable, in my opinion, with native code and OCI images (which are system specific regardless), and we do preserve that capability when using different dependencies by system (which we have regardless, because some of our own dependencies are organized as different jars for each architecture/system)

This suggestion would match a common pattern in open source, but you are correct that each approach has different benefits and costs.